codec/decode.go

```diff
@@ -2068,8 +2068,8 @@ func (d *Decoder) deferred(err1 *error) {
 	d.alwaysAtEnd()
 	if recoverPanicToErr {
 		if x := recover(); x != nil {
-			panicValToErr(d, x, err1)
-			panicValToErr(d, x, &d.err)
+			*err1 = panicValToErr(d, x)
+			d.err = *err1
```

Comment:
Seems better here to avoid calling panicValToErr (and the reflection calls) twice.